src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -304,6 +308,101 @@ CreateDistributedInsertSelectPlan(Query *originalQuery,
 }
 
 
+/*
+ * CreateInsertSelectIntoLocalTablePlan
+ */
+DistributedPlan *
+CreateInsertSelectIntoLocalTablePlan(uint64 planId, Query *originalQuery, ParamListInfo
+									 boundParams, bool hasUnresolvedParams,
+									 PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RangeTblEntry *selectRte = ExtractSelectRangeTableEntry(originalQuery);
+
+	/*RangeTblEntry *insertRte = ExtractResultRelationRTE(originalQuery); */
+	/*Oid targetRelationId = insertRte->relid; */
+	Query *selectQuery = selectRte->subquery;
+	DistributedPlan *distPlan = CreateDistributedPlan(planId, selectQuery, selectQuery,
+													  boundParams, hasUnresolvedParams,
+													  plannerRestrictionContext);
+	if (distPlan->masterQuery == NULL)
+	{
+		distPlan->masterQuery = CreateMasterQueryForRouterInsertSelectIntoLocalTable(
+			distPlan);
+	}
+
+	/* masterQuery of a distributed select is for combining the results from
+	 * worker nodes on the coordinator node. Putting it as a subquery to the
+	 * insert query makes insert query insert the combined select value from
+	 * the workers. And making the resulting insert query the masterQuery let's
+	 * us execute this insert command.
+	 *
+	 * So this operation makes the master query insert the result of the
+	 * distributed select instead of returning it.
+	 */
+	selectRte->subquery = distPlan->masterQuery;
+	distPlan->masterQuery = originalQuery;
+
+	return distPlan;
+}
+
+
+/*
+ * CreateMasterQueryForRouterInsertSelectIntoLocalTable
+ */
+static Query *
+CreateMasterQueryForRouterInsertSelectIntoLocalTable(DistributedPlan *distPlan)
```

Comment:
Maybe CreateMasterQueryForRouterInsertSelectIntoLocalTable -> CreateMasterQueryForRouterPlan, since it's not that specific to insert..select.